modules/DNN/RBM.cpp

```diff
@@ -0,0 +1,121 @@
+#include "RBM.h"
+#include "DGM/random.h"
+#include "macroses.h"
+
+namespace DirectGraphicalModels {
+	namespace dnn {
+		CRBM::CRBM(const std::vector<ptr_nl_t>& vpLayers){
+			for (auto& nl : vpLayers)
+				m_vpNeuronLayers.push_back(nl);
+		}
+
+		Mat CRBM::getBinomial(Mat mean) {
+			Mat res(mean.clone());
+			for (int y = 0; y < res.rows; y++) {
+				float* pRes = res.ptr<float>(y);
+				for (int x = 0; x < res.cols; x++) {
+					if (pRes[x] < 0 || pRes[x]>1) {
+						pRes[x] = 0;
+					}
+					double r = random::U<double>();	// uniformly distributed random number betwee 0 and 1
+					if (r < pRes[x])
+					{
+						pRes[x] = 1;
+					}
+					else
+					{
+						pRes[x] = 0;
+					}
+				}
+			}
+			return res;
+		}
+
+		void CRBM::debug() {
+			std::cout << "Weight - rows: " << m_vpNeuronLayers[1]->getWeights().rows << " cols: " << m_vpNeuronLayers[1]->getWeights().cols << std::endl;
+
+			std::cout << "Positive H mean - rows: " << m_positiveHMean.rows << " cols: " << m_positiveHMean.cols << std::endl;
+			std::cout << "Positive H sample - rows: " << m_positiveHSample.rows << " cols: " << m_positiveHSample.cols << std::endl;
+			std::cout << "Negative H mean - rows: " << m_negativeHMean.rows << " cols: " << m_negativeHMean.cols << std::endl;
+			std::cout << "Negative H sample - rows: " << m_negativeHSample.rows << " cols: " << m_negativeHSample.cols << std::endl;
+			std::cout << "Negative V mean - rows: " << m_negativeVMean.rows << " cols: " << m_negativeVMean.cols << std::endl;
+			std::cout << "Negative V sample - rows: " << m_negativeVSample.rows << " cols: " << m_negativeVSample.cols << std::endl;    
+		}
+
+		void CRBM::sampleVisible(Mat values) {
+			m_negativeVMean = propagateDown(values);
+			m_negativeVSample = getBinomial(m_negativeVMean);
+		}
+
+		void CRBM::sampleHiddenPositive(Mat values) {
+			m_positiveHMean = propagateUp(values);		
+			m_positiveHSample = getBinomial(m_positiveHMean);
+
+			/*for (int y = 0; y < sample.rows; y++) {
+				float* pRess = sample.ptr<float>(y);
+				for (int x = 0; x < sample.cols; x++)
+					std::cout << pRess[x] << std::endl;
+			}*/
+		}
+
+		void CRBM::sampleHiddenNegative(Mat values) {
+			m_negativeHMean = propagateUp(values);
+			m_negativeHSample = getBinomial(m_negativeHMean);
+		}
+
+		Mat CRBM::propagateUp(Mat values) {
+			m_vpNeuronLayers[0]->setNetValues(values); //set the visible layer values
+
+			m_vpNeuronLayers[1]->dotProd(m_vpNeuronLayers[0]->getValues()); //sigmoid(sum(visible * weights)+bias)
+
+			return m_vpNeuronLayers.back()->getValues();
+		}
+
+		Mat CRBM::propagateDown(Mat values){
+			m_vpNeuronLayers[0]->dotProdVis(values, m_vpNeuronLayers[1]->getWeights()); 

```

Comment:
Can we use here just the old` dotProd()` method?
_I.e._` m_vpNeuronLayers[1]->dotProd(values); ` (index 1 instead of 0) ?

Reply:
I have tried my best to explain why I used `dotProdVis(values, weights)` above. Please check it out.